Mathlib/Data/Multiset/Bind.lean

```diff
@@ -0,0 +1,361 @@
+#align multiset.bind_assoc Multiset.bind_assoc
+
+theorem bind_bind (m : Multiset α) (n : Multiset β) {f : α → β → Multiset γ} :
+    ((bind m) fun a => (bind n) fun b => f a b) = (bind n) fun b => (bind m) fun a => f a b :=
+  Multiset.induction_on m (by simp) (by simp (config := { contextual := true }))
+#align multiset.bind_bind Multiset.bind_bind
+
+theorem bind_map_comm (m : Multiset α) (n : Multiset β) {f : α → β → γ} :
+    ((bind m) fun a => n.map fun b => f a b) = (bind n) fun b => m.map fun a => f a b :=
+  Multiset.induction_on m (by simp) (by simp (config := { contextual := true }))
+#align multiset.bind_map_comm Multiset.bind_map_comm
+
+@[to_additive (attr := simp)]
+theorem prod_bind [CommMonoid β] (s : Multiset α) (t : α → Multiset β) :
+    (s.bind t).prod = (s.map fun a => (t a).prod).prod :=
+  Multiset.induction_on s (by simp) fun a s ih => by simp [ih, cons_bind]
+#align multiset.prod_bind Multiset.prod_bind
+
+theorem rel_bind {r : α → β → Prop} {p : γ → δ → Prop} {s t} {f : α → Multiset γ}
+    {g : β → Multiset δ} (h : (r ⇒ Rel p) f g) (hst : Rel r s t) : Rel p (s.bind f) (t.bind g) :=
+  by
+  apply rel_join
+  rw [rel_map]
+  exact hst.mono fun a _ b _ hr => h hr
+#align multiset.rel_bind Multiset.rel_bind
+
+theorem count_sum [DecidableEq α] {m : Multiset β} {f : β → Multiset α} {a : α} :
+    count a (map f m).sum = sum (m.map fun b => count a <| f b) :=
+  Multiset.induction_on m (by simp) (by simp)
+#align multiset.count_sum Multiset.count_sum
+
+theorem count_bind [DecidableEq α] {m : Multiset β} {f : β → Multiset α} {a : α} :
+    count a (bind m f) = sum (m.map fun b => count a <| f b) :=
+  count_sum
+#align multiset.count_bind Multiset.count_bind
+
+theorem le_bind {α β : Type _} {f : α → Multiset β} (S : Multiset α) {x : α} (hx : x ∈ S) :
+    f x ≤ S.bind f := by
+  classical
+    rw [le_iff_count]
+    intro a
+    rw [count_bind]
+    apply le_sum_of_mem
+    rw [mem_map]
+    exact ⟨x, hx, rfl⟩
+#align multiset.le_bind Multiset.le_bind
+
+-- Porting note: @[simp] removed because not in normal form
+theorem attach_bind_coe (s : Multiset α) (f : α → Multiset β) :
+    (s.attach.bind fun i => f i) = s.bind f :=
+  congr_arg join <| attach_map_val' _ _
+#align multiset.attach_bind_coe Multiset.attach_bind_coe
+
+end Bind
+
+/-! ### Product of two multisets -/
+
+
+section Product
+
+variable (a : α) (b : β) (s : Multiset α) (t : Multiset β)
+
+/-- The multiplicity of `(a, b)` in `s ×ˢ t` is
+  the product of the multiplicity of `a` in `s` and `b` in `t`. -/
+def product (s : Multiset α) (t : Multiset β) : Multiset (α × β) :=
+  s.bind fun a => t.map <| Prod.mk a
+#align multiset.product Multiset.product
+
+-- This notation binds more strongly than (pre)images, unions and intersections.
+@[inherit_doc]
+infixr:82 " ×ˢ " => Multiset.product
+
+@[simp]
+theorem coe_product (l₁ : List α) (l₂ : List β) : @product α β l₁ l₂ = l₁.product l₂ :=
+  by
```

Comment:
```suggestion
theorem coe_product (l₁ : List α) (l₂ : List β) : @product α β l₁ l₂ = l₁.product l₂ := by
```
Maybe run the script from #1525 or so? It doesn't seem to take a filename atm...